File: solutions/golang/parking-lot/parkingLot.go
package main

import (
	"errors"
	"fmt"
	"github.com/ashishps1/awesome-low-level-design/solutions/golang/parking-lot/vehicles"
	"sync"
)

var (
	singletonMap map[int]*ParkingLot
	once         sync.Once
	instanceLock sync.Mutex
	parkLock     sync.Mutex
)

type ParkingLot struct {
	Name    string
	Address string // this could be another struct but keeping it simple for now
	Levels  []*Level
}

func GetInstance(id int) *ParkingLot {
	once.Do(func() {
		singletonMap = make(map[int]*ParkingLot)
	})
	if _, isPresent := singletonMap[id]; !isPresent {
		instanceLock.Lock()
		if _, isPresent = singletonMap[id]; !isPresent {
			fmt.Printf("Creating new parking lot for id: %d\n", id)
			pLot := ParkingLot{}
			singletonMap[id] = &pLot
		}
		instanceLock.Unlock()
	}
	return singletonMap[id]
}

func (p *ParkingLot) ParkVehicle(vehicle vehicles.Vehicle) (*Ticket, error) {
	// Find the best spot available
	parkLock.Lock()
	defer parkLock.Unlock()

	fmt.Printf("Parking vehicle no: %s\n", vehicle.RegisNo())
	var spot *ParkingSpot
	for _, level := range p.Levels {
		if spot = level.FindParkingSpot(vehicle); spot != nil {
			break
		} else {
			return nil, errors.New("parking full")
		}
	}
	// Mark spot as occupied
	spot.CurrentVehicle = vehicle
	// Generate ticket and return
	ticket := NewTicket(spot, vehicle)

	return ticket, nil
}

func (p *ParkingLot) DisplayAvailability() {
	fmt.Println("Following no of spots available:")
	for floor, level := range p.Levels {
		fmt.Printf("level %d:\n", floor)
		level.DisplayAvailability()
		fmt.Println()
	}
	fmt.Println()
}

func (p *ParkingLot) UnParkVehicle(ticket *Ticket) {
	// empty the spot
	ticket.ParkingSpot.CurrentVehicle = nil
	// calculate the amount
	ticket.CalculateAmount()
}

func (p *ParkingLot) AddLevel(floor int) {
	p.Levels = append(p.Levels, NewLevel(floor))
}

File: solutions/my-javascript/parkinglotsystem/ParkingLot.js
export class ParkingLot {
    constructor() {
        this.instance = null;
        this.levels = [];
    }

    getInstance() {
        if (this.instance === null) {
            this.instance = new ParkingLot();
        }
        return this.instance;
    }

    addLevel(level) {
        this.levels.push(level);
    }

    // traverse to find available parking spot
    // park the vehicle in the first available spot
    parkVehicle(vehicle) {
        for (const level of this.levels) {
            const parkedSpot = level.parkVehicle(vehicle);
            if (parkedSpot) {
                console.log(`Vehicle parked in ${parkedSpot.spotNumber}`);
                return true;
            }
        }
        console.log(`No available parking spots for vehicle`);
        return false;
    }

    // traverse to find the vehicle
    // unpark the vehicle from the spot
    unParkVehicle(vehicle) {
        for (const level of this.levels) {
            if (level.unParkVehicle(vehicle)) {
                console.log(`Vehicle unparked from`)
                return true;
            }
        }
        console.log(`Vehicle not found`);
        return false;
    }

    displayAvailableSpots() {
        console.log(`Available parking spots:`);
        for (let i = 0; i < this.levels.length; i++) {
            this.levels[i].displayAvailableSpots();
        }
    }
}